Warm up the query string obfuscator regex

kevingosse opened the pull request:
## Summary of changes

It looks like the regex is compiled during the first invocation, not during construction, which might explain the timeouts we're seeing. Now warmup the regex in the constructor.

## Reason for change

Timeouts occur in tests, and telemetry show that they occur in the wild too.

See also: https://github.com/dotnet/runtime/issues/54747

## Implementation details

Matching `o` in the obfuscator constructor, and catching all exceptions. I initially tried with an empty string, but it looks like some values are optimized and short-circuit the regex.

andrewlock commented on tracer/test/Datadog.Trace.ClrProfiler.IntegrationTests/GraphQLTests.cs:
```diff
@@ -192,10 +192,6 @@ protected GraphQLTests(string sampleAppName, ITestOutputHelper output, string te
             SetServiceVersion(ServiceVersion);
             SetEnvironmentVariable("DD_TRACE_SPAN_ATTRIBUTE_SCHEMA", metadataSchemaVersion);
 
-            // After updating the regex implementation multiple runtimes started hitting the timeout.
-            // Increasing timeout here to prevent flaky tests.
-            SetEnvironmentVariable(Configuration.ConfigurationKeys.ObfuscationQueryStringRegexTimeout, "5000");
-
```
Bold move

andrewlock commented on tracer/src/Datadog.Trace/Util/Http/QueryStringObfuscation/Obfuscator.cs:
```diff
@@ -29,6 +26,18 @@ internal Obfuscator(string pattern, TimeSpan timeout, IDatadogLogger logger)
+            catch (Exception exception)
+            {
+                // We're in the constructor, we can't let exceptions escape
+                _logger.Warning(exception, "Timeout during the query string obfuscation regex warmup, with a timeout value of {TotalMilliseconds} ms and regex pattern {Pattern}", _timeout.TotalMilliseconds, _regex.ToString());
```
Do we actually care about timeouts during compilation, assuming that even if it does timeout, the compilation completes? 🤔

kevingosse replied:
We don't care much, that's why I lowered to warning

andrewlock replied:
We don't ever bail-out of the compilation, even if it times out, right?

kevingosse replied:
ok ok, removing the log 😄

andrewlock reviewed (approved):
I mean, I would personally be inclined to keep the timeout boosted (although _everywhere_) but I guess we'll see if I can say "I told you so"

andrewlock commented:
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5266)](https://github.com/DataDog/dd-trace-dotnet/tree/281d3eeb2a15161f2eafa5cf30f70021372f05b9)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/21558a991a89bc7bef07514335faf62422fbc86e)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5266) - mean (73ms)  : 66, 80
     .   : milestone, 73,
    master - mean (74ms)  : 66, 82
     .   : milestone, 74,

    section CallTarget+Inlining+NGEN
    This PR (5266) - mean (980ms)  : 956, 1005
     .   : milestone, 980,
    master - mean (995ms)  : 967, 1024
     .   : milestone, 995,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5266) - mean (110ms)  : 107, 114
     .   : milestone, 110,
    master - mean (111ms)  : 107, 115
     .   : milestone, 111,

    section CallTarget+Inlining+NGEN
    This PR (5266) - mean (712ms)  : 688, 735
     .   : milestone, 712,
    master - mean (716ms)  : 695, 736
     .   : milestone, 716,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5266) - mean (93ms)  : 90, 97
     .   : milestone, 93,
    master - mean (94ms)  : 91, 97
     .   : milestone, 94,

    section CallTarget+Inlining+NGEN
    This PR (5266) - mean (668ms)  : 641, 695
     .   : milestone, 668,
    master - mean (675ms)  : 648, 701
     .   : milestone, 675,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5266) - mean (187ms)  : 184, 191
     .   : milestone, 187,
    master - mean (188ms)  : 183, 192
     .   : milestone, 188,

    section CallTarget+Inlining+NGEN
    This PR (5266) - mean (1,059ms)  : 1034, 1084
     .   : milestone, 1059,
    master - mean (1,050ms)  : 1026, 1074
     .   : milestone, 1050,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5266) - mean (270ms)  : 262, 278
     .   : milestone, 270,
    master - mean (269ms)  : 262, 275
     .   : milestone, 269,

    section CallTarget+Inlining+NGEN
    This PR (5266) - mean (866ms)  : 841, 891
     .   : milestone, 866,
    master - mean (860ms)  : 829, 891
     .   : milestone, 860,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5266) - mean (261ms)  : 256, 265
     .   : milestone, 261,
    master - mean (257ms)  : 252, 263
     .   : milestone, 257,

    section CallTarget+Inlining+NGEN
    This PR (5266) - mean (854ms)  : 832, 877
     .   : milestone, 854,
    master - mean (842ms)  : 817, 867
     .   : milestone, 842,

```

andrewlock commented:
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5266)](https://github.com/DataDog/dd-trace-dotnet/tree/281d3eeb2a15161f2eafa5cf30f70021372f05b9)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/21558a991a89bc7bef07514335faf62422fbc86e)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5266) (11.027M)   : 0, 11027168
    master (10.925M)   : 0, 10925303
    benchmarks/2.9.0 (10.620M)   : 0, 10620397

    section Automatic
    This PR (5266) (7.676M)   : 0, 7675574
    master (7.496M)   : 0, 7496085
    benchmarks/2.9.0 (7.702M)   : 0, 7701617

    section Trace stats
    This PR (5266) (7.888M)   : 0, 7887981
    master (7.872M)   : 0, 7871947

    section Manual
    This PR (5266) (9.607M)   : 0, 9607231
    master (9.635M)   : 0, 9635035

    section Manual + Automatic
    This PR (5266) (7.272M)   : 0, 7271867
    master (7.114M)   : 0, 7113688

    section Version Conflict
    This PR (5266) (6.409M)   : 0, 6409175
    master (6.292M)   : 0, 6292435

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5266) (9.629M)   : 0, 9629288
    benchmarks/2.9.0 (9.526M)   : 0, 9525834

    section Automatic
    This PR (5266) (6.694M)   : 0, 6693837

    section Trace stats
    This PR (5266) (6.857M)   : 0, 6856945

    section Manual
    This PR (5266) (8.539M)   : 0, 8538647

    section Manual + Automatic
    This PR (5266) (6.198M)   : 0, 6198182

    section Version Conflict
    This PR (5266) (5.725M)   : 0, 5725422

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5266) (10.235M)   : 0, 10235203
    master (9.856M)   : 0, 9856447
    benchmarks/2.9.0 (10.533M)   : 0, 10532864

    section Automatic
    This PR (5266) (7.487M)   : 0, 7487139
    master (7.076M)   : 0, 7076232
    benchmarks/2.9.0 (7.815M)   : 0, 7814762

    section Trace stats
    This PR (5266) (7.597M)   : 0, 7596712
    master (7.404M)   : 0, 7404111

    section Manual
    This PR (5266) (9.349M)   : 0, 9348947
    master (8.658M)   : 0, 8657799

    section Manual + Automatic
    This PR (5266) (7.193M)   : 0, 7192785
    master (6.832M)   : 0, 6831716

    section Version Conflict
    This PR (5266) (6.348M)   : 0, 6348073
    master (6.173M)   : 0, 6172929

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    master (7.559M)   : 0, 7559179
    benchmarks/2.9.0 (7.859M)   : 0, 7859139

    section No attack
    master (1.844M)   : 0, 1844300
    benchmarks/2.9.0 (3.196M)   : 0, 3196117

    section Attack
    master (1.451M)   : 0, 1451050
    benchmarks/2.9.0 (2.519M)   : 0, 2519149

    section Blocking
    master (3.102M)   : 0, 3101736

    section IAST default
    master (6.434M)   : 0, 6433840

    section IAST full
    master (5.626M)   : 0, 5625514

    section Base vuln
    master (0.917M)   : 0, 917115

    section IAST vuln
    master (0.881M)   : 0, 881291

```

lucaspimentel commented on tracer/src/Datadog.Trace/Util/Http/QueryStringObfuscation/Obfuscator.cs:
```diff
@@ -29,6 +26,17 @@ internal Obfuscator(string pattern, TimeSpan timeout, IDatadogLogger logger)
+            {
+                // Warmup the regex
+                // Can't use empty string, space, or dot, as they are optimized and don't actually trigger the compilation
+                _ = _regex.Match("o");
```
Is this good enough? The `Obfuscator` itself isn't constructed until something calls `QueryStringManager.TruncateAndObfuscate()` for the first time because it uses a `Lazy<ObfuscatorBase>`.

lucaspimentel replied:
Maybe on the next line after this? https://github.com/DataDog/dd-trace-dotnet/blob/2fe16fa0a29c9b511dcab1da40d77348682292a4/tracer/src/Datadog.Trace/TracerManager.cs#L86

```csharp
QueryStringManager = new(...);

if (settings.QueryStringReportingEnabled)
{
    QueryStringManager.TruncateAndObfuscate("o");
}
```

lucaspimentel replied:
This makes me wonder if we need the `Lazy<ObfuscatorBase>` anymore. We will now initialize the obfuscator eagerly or not at all.

kevingosse replied:
The goal was really to reduce the number of timeouts (that gets reported through telemetry and make the tests flaky) rather than moving the initialization out of the first request. Eagerly building the regex seems much more complex (it's not a huge deal but we might end up building up for applications that don't need it, which is not great)

lucaspimentel replied:
Oops, I misunderstood the goal, then. Thanks for clarifying!

lucaspimentel reviewed (changes requested):
I don't think this initializes the regex the way we want to.

lucaspimentel commented:
PR description:
> I reverted some of the timeout changes introduced in https://github.com/DataDog/dd-trace-dotnet/pull/5255 because I'm hopelessly optimistic.

I don't see these changes.

kevingosse commented:
> PR description:
> 
> > I reverted some of the timeout changes introduced in #5255 because I'm hopelessly optimistic.
> 
> I don't see these changes.

Right. I reverted the revert because it was indeed too optimistic. Fixed the description.
